packages/multi-test/src/app.rs

```diff
@@ -269,65 +312,107 @@ where
         }
     }
 
+    pub fn execute(&mut self, sender: Addr, msg: CosmosMsg<C>) -> Result<AppResponse, String> {
+        match msg {
+            CosmosMsg::Wasm(msg) => {
+                let (resender, res) = self.execute_wasm(sender, msg)?;
+                self.process_response(resender, res)
+            }
+            CosmosMsg::Bank(msg) => {
+                self.bank.execute(sender, msg)?;
+                Ok(AppResponse::default())
+            }
+            _ => unimplemented!(),
+        }
+    }
+
     /// This will execute the given messages, making all changes to the local cache.
     /// This *will* write some data to the cache if the message fails half-way through.
     /// All sequential calls to RouterCache will be one atomic unit (all commit or all fail).
     ///
     /// For normal use cases, you can use Router::execute() or Router::execute_multi().
     /// This is designed to be handled internally as part of larger process flows.
-    fn execute(&mut self, sender: Addr, msg: SubMsg<C>) -> Result<AppResponse, String> {
-        // TODO: actually handle reply semantics
-        if msg.reply_on != ReplyOn::Never {
-            unimplemented!();
+    fn execute_submsg(&mut self, contract: Addr, msg: SubMsg<C>) -> Result<AppResponse, String> {
+        // execute in cache
+        let mut subtx = self.cache();
+        let res = subtx.execute(contract.clone(), msg.msg);
+        if res.is_ok() {
+            subtx.prepare().commit(self);
         }
-        match msg.msg {
-            CosmosMsg::Wasm(msg) => {
-                let (resender, res) = self.execute_wasm(sender, msg)?;
-                let mut attributes = res.attributes;
-                let mut events = res.events;
-                // recurse in all messages
-                for resend in res.messages {
-                    let subres = self.execute(resender.clone(), resend)?;
-                    // ignore the data now, just like in wasmd
-                    // append the events
-                    // TODO: is this correct for attributes??? Or do we turn them into sub-events?
-                    attributes.extend_from_slice(&subres.attributes);
-                    events.extend_from_slice(&subres.events);
+
+        // call reply if meaningful
+        if let Ok(r) = res {
+            if matches!(msg.reply_on, ReplyOn::Always | ReplyOn::Success) {
+                let mut orig = r.clone();
+                let reply = Reply {
+                    id: msg.id,
+                    result: ContractResult::Ok(SubMsgExecutionResponse {
+                        events: r.events,
+                        data: r.data,
```

Comment:
For the better reply test, maybe we can wrap the data with Protobuf wrapper like 

https://github.com/terraswap/terraswap/blob/55c9ac2e54e95462c5c2ed3507cbdc27993b1b64/contracts/terraswap_factory/src/response.proto#L4-L9

https://github.com/CosmWasm/wasmd/blob/8bec411e4bd947cc7d872bd14bd1da865bf703bb/proto/cosmwasm/wasm/v1beta1/tx.proto#L69-L74

Then, we can properly test Protobuf decoding part at reply like https://github.com/terraswap/terraswap/blob/55c9ac2e54e95462c5c2ed3507cbdc27993b1b64/contracts/terraswap_factory/src/contract.rs#L152-L155

Reply:
Nice catch. I did something in `init_response`, but that was the old (pre-protobuf) version. Thanks for pointing this out.

Reply:
Added #330 for this

Reply:
And closing with #331